### PR TITLE
NOX-127 [Nox.Generators] Entity with multiple relations to the same entity - improve relation discover

### PR DESCRIPTION
**Pull request by emir-sehovic1**
…ations to same other entity

### CODE REVIEW

File: src/Nox.Solution/Models/Domain/Entities/Relationships/EntityRelationship.cs [hunk 1]
```diff
@@ -39,6 +39,9 @@ public class EntityRelationship : YamlConfigNode<NoxSolution,Entity>
     )]
     public string Entity { get; internal set; } = null!;
 
+    [Description("The name of the relationship in the related entity that this relationship refers to.")]
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Relationships/EntityRelationship.cs]:
```suggestion
    [Title("The relationship name in the related entity that this relationship refers to.")]
    [Description("Required when having multiple relations to the same entity allowing to unambiguous find the correct association.")]
```

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 2]
```diff
@@ -416,7 +416,68 @@ private void ValidateThatAllRelatedEntitiesHaveSimpleKeys(ValidationResult resul
...
+        {
+            result.Errors.Add(new ValidationFailure(nameof(Name), message));
+        }
+
+        // Validate that in case there are multiple relationships to the same entity, that the RefRelationshipName is populated
+        var messages2 = relationshipsGroupedByRelatedEntity
+            .Where(g => g.Count() > 1)
+            .SelectMany(g => g, (g, r) => new { RelatedEntity = g.Key, Relationship = r})
+            .Where(x => string.IsNullOrWhiteSpace(x.Relationship.RefRelationshipName))
+            .Select(x => $"Relationship [{x.Relationship.Name}] on entity [{Name}] is missing RefRelationshipName value, but there are multiple relationships that refer to [{x.RelatedEntity}].");
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs]:
```suggestion
            .Select(x => $"The relationship name [{x.Relationship.Name}] on the entity [{Name}] lacks RefRelationshipName value. With multiple relationships referencing  [{x.RelatedEntity}], is not possible to unambiguously select the correct association.");
```

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 3]
```diff
@@ -416,7 +416,68 @@ private void ValidateThatAllRelatedEntitiesHaveSimpleKeys(ValidationResult resul
...
+    }
+
+    private void ValidateRefRelationshipNames(ValidationResult result)
+    {
+        var relationshipsGroupedByRelatedEntity = Relationships.GroupBy(x => x.Entity);
+
+        // Validate that in case there are no multiple relationships to the same entity, that the RefRelationshipName is not populated
+        var messages1 = relationshipsGroupedByRelatedEntity
+            .Where(g => g.Count() == 1 && !string.IsNullOrWhiteSpace(g.First().RefRelationshipName))
+            .Select(x => $"Relationship [{x.First().Name}] on entity [{Name}] has a non-null RefRelationshipName value, but there is only one relationship that refers to [{x.First().Entity}].");
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs]:
```suggestion
            .Select(x => $"The relationship name [{x.First().Name}] on the entity [{Name}] has a non-null RefRelationshipName value. There is only one relationship referencing [{x.First().Entity}], RefRelationshipName is unnecessary in this cases.");
```

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 4]
```diff
@@ -169,12 +170,10 @@ public void SetDefaultOtherEntityOnRelationships(Domain parentNode)
...
             var otherEntity = grouping.Key;
             var relatedRelationships = otherEntity.Relationships.Where(r => r.Entity.Equals(Name)).ToList();
-            var count = Math.Min(grouping.Count(), relatedRelationships.Count);
-            for (var i = 0; i < count; i++)
+            foreach(var relationship in grouping)
             {
-                var relationship = grouping.ElementAt(i);
                 relationship.Related.Entity = otherEntity;
-                relationship.Related.EntityRelationship = relatedRelationships[i];
+                relationship.Related.EntityRelationship = relatedRelationships.Find(x => x.Name == relationship.RefRelationshipName) ?? relatedRelationships[0];
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs]: for performancing improvement, maybe we can do this?:
```suggestion
                relationship.Related.EntityRelationship = relatedRelationships.Find(x => relationship.RefRelationshipName is not null &&  x.Name == relationship.RefRelationshipName) ?? relatedRelationships[0];
```

**Review by rochar — COMMENTED** [review]
looks good, just a couple of suggestions

**Comment by emir-sehovic1** [comment]
Thank you @rochar , the suggestions were applied.

**Review by rochar — APPROVED** [review]

**Review by gjorgji-grgovski-authoritypartners — APPROVED** [review]

**Pull request merged**